### PR TITLE
test(wallet-integration): unit tests for pop up contract

### PR DESCRIPTION
**Pull request by al3mart**
This PR covers the addition of unit tests for the wallet integration with `pop up contracts`.

~~`fn wait_for_signature` is tested as part of an integration test in `pop-cli/tests/contracts.rs`.~~
The integration test the https server handling the signature of payloads has been included in the test `contract_lifecycle` in `pop-cli/tests/contracts.rs`.

Retrieving the correct upload and instantiation call data is tested in `pop-cli/src/commands/up/contract.rs` in `get_upload_call_data_works` and `get_instantiate_call_data_works` tests.

[sc-1887]

### CODE REVIEW

File: crates/pop-cli/src/commands/up/contract.rs [hunk 1]
```diff
@@ -438,15 +470,109 @@ mod tests {
...
+		let localhost_url = format!("ws://127.0.0.1:{}", port);
+		sleep(Duration::from_secs(20)).await;
+
+		let up_contract_opts = UpContractCommand {
+			path: Some(temp_dir.path().join("testing")),
+			constructor: "new".to_string(),
+			args: vec!["false".to_string()],
+			value: "0".to_string(),
+			gas_limit: None,
+			proof_size: None,
```
Comment (peterwht) [inline crates/pop-cli/src/commands/up/contract.rs:542]: would be nice if these were some value to ensure pop-cli isn't just taking the default None.

File: crates/pop-cli/src/common/wallet.rs [hunk 2]
```diff
@@ -46,6 +46,7 @@ mod tests {
 	// This is a helper test for an actual running pop CLI.
 	// It can serve as the "frontend" to query the payload, sign it
 	// and submit back to the CLI.
+	#[ignore]
 	#[tokio::test]
 	async fn sign_call_data() -> anyhow::Result<()> {
```
Comment (peterwht) [inline crates/pop-cli/src/common/wallet.rs:51]: can be removed if its purpose is properly represented elsewhere in the tests.

File: crates/pop-cli/tests/contract.rs [hunk 3]
```diff
@@ -7,14 +7,52 @@ use pop_contracts::{
...
 use strum::VariantArray;
+use subxt::{config::DefaultExtrinsicParamsBuilder as Params, tx::Payload, utils::to_hex};
+use subxt_signer::sr25519::dev;
+use tokio::time::sleep;
 use url::Url;
 
+// This struct implements the [`Payload`] trait and is used to submit
+// pre-encoded SCALE call data directly, without the dynamic construction of transactions.
+struct CallData(Vec<u8>);
+impl Payload for CallData {
```
Comment (peterwht) [inline crates/pop-cli/tests/contract.rs:21]: leaving a note: I copied this from the pop call parachain PR (IIRC). Once everything gets merged in we should hopefully be able to just import this.

File: crates/pop-cli/tests/contract.rs [hunk 4]
```diff
@@ -7,14 +7,52 @@ use pop_contracts::{
...
+		out.extend_from_slice(&self.0);
+		Ok(())
+	}
+}
+
+// TransactionData has been copied from wallet_integration.rs
+/// Transaction payload to be sent to frontend for signing.
+#[derive(Serialize, Debug)]
+#[cfg_attr(test, derive(Deserialize, Clone))]
+pub struct TransactionData {
```
Comment (peterwht) [inline crates/pop-cli/tests/contract.rs:36]: any reason this wasn't just imported?

File: crates/pop-cli/tests/contract.rs [hunk 5]
```diff
@@ -7,14 +7,52 @@ use pop_contracts::{
...
+	pub fn call_data(&self) -> Vec<u8> {
+		self.call_data.clone()
+	}
+}
+
 /// Test the contract lifecycle: new, build, up, call
 #[tokio::test]
 async fn contract_lifecycle() -> Result<()> {
 	const DEFAULT_PORT: u16 = 9944;
+	const DEFAULT_ENDPOINT: &str = "ws://127.0.0.1:9944";
```
Comment (peterwht) [inline crates/pop-cli/tests/contract.rs:53]: would recommend changing this port away from 9944 -- maybe use free_port as well?

File: crates/pop-cli/tests/contract.rs [hunk 6]
```diff
@@ -7,14 +7,52 @@ use pop_contracts::{
...
+	const WALLET_INT_URI: &str = "http://127.0.0.1:9090";
```
Comment (peterwht) [inline crates/pop-cli/tests/contract.rs:54]: Not for this PR, but likely need something to componentize things a little more and let us alter the address for the wallet integration. This port 9090 is going to conflict with some other tests.

**Review by peterwht — COMMENTED** [review]
Nice!!! Looking good. Left a few comments for minor adjustments / thoughts.

File: crates/pop-cli/tests/contract.rs (hunk 6 above)
Comment (al3mart) [reply crates/pop-cli/tests/contract.rs:54]: Agreed!

File: crates/pop-cli/src/common/wallet.rs (hunk 2 above)
Comment (al3mart) [reply crates/pop-cli/src/common/wallet.rs:51]: Makes sense. Even more now that we have the front end part too.

File: crates/pop-cli/src/commands/up/contract.rs (hunk 1 above)
Comment (al3mart) [reply crates/pop-cli/src/commands/up/contract.rs]: Addressed in [86ab4b8](https://github.com/r0gue-io/pop-cli/pull/374/commits/86ab4b88eadfd6955982e12ca849fe0bcd2bab55)

File: crates/pop-cli/src/common/wallet.rs (hunk 2 above)
Comment (al3mart) [reply crates/pop-cli/src/common/wallet.rs]: Removed in [f9851d1](https://github.com/r0gue-io/pop-cli/pull/374/commits/f9851d1ec0bd761e83f0f4ab833dec344b9b5ba2)

File: crates/pop-cli/tests/contract.rs (hunk 5 above)
Comment (al3mart) [reply crates/pop-cli/tests/contract.rs]: Addressed in [80f4f7b](https://github.com/r0gue-io/pop-cli/pull/374/commits/80f4f7bd79f5a6b3f0f64b0baea7f4e840b6d7ca)

**Pull request merged**